### DISCUSSION TITLE
Row level security with select statments

### DISCUSSION THREAD

**Discussion by Nick-Mazuk** [supabase/supabase#1669] [Questions] [answered]
I'm trying to restrict certain read/write operations to users with custom roles (e.g., admin, editor, author, etc.) can create, edit, or delete rows in the `posts` collection. How can I do this with row level security?

Here's what I've tried so far. I have a table created for each of these roles…

```sql
create table "userRoles" (
  uid uuid primary key references users(uid),
  "isAdmin" boolean default false not null,
  "isAuthor" boolean default false not null,
  "isEditor" boolean default false not null
);
```

…where the `users` table is a [public.users](https://supabase.io/docs/guides/auth#create-a-publicusers-table) table. And another for posts

```sql
create table posts (
  ...
)
```

Then in the Supabase dashboard, I tried using the following rules individually on the `posts` table, but I kept getting 403 errors:

```sql
(uid() in ( select uid from "userRoles" where isAdmin = true))
```
```sql
(uid() in ( select "userRoles".uid from "userRoles" where "userRoles".isAdmin = true))
```
```sql
(uid() in ( select uid from "userRoles" where isAdmin = true and uid = uid()))
```
```sql
(exists ( select 1 from "userRoles" where isAdmin = true and uid = uid()))
```

Any ideas where I'm going wrong?

_P.S. I'm willing to completely restructure how I'm saving these roles if there's a better way._

**Comment by soedirgo** [comment]
Are you running this in NodeJS? There was an issue on storing the logged in user in NodeJS: https://github.com/supabase/gotrue-js/issues/34. The workaround is to run `supabase.auth.setAuth()` using the `access_token`.

Btw, you can simplify your policy to `auth.uid() = uid and "isAdmin" = true`.

**Reply by Nick-Mazuk** [reply]
I'm running this is the browser, not NodeJS. And I'm fairly certain the error is from the RLS because changing it to `uid() = uid` works. I have used the new `supabase.auth.setAuth(token)` api in Node and it's awesome, though!

I should clarify that the rule is not for the `userRoles` table. It's for the `posts` table. I'm looking to only let certain people create, edit, or delete posts. So I'd assume the simplified policy you suggested won't work since `isAdmin` isn't in the `posts` table. Updating the original post with that clarification.

**Reply by soedirgo** [reply]
OK, I think I got it - by default, when doing e.g. an insert, the client does both `INSERT` *and* `SELECT` (on the inserted rows). So you also need a policy to allow the admins to `SELECT`.

Alternatively, for write-only rows, you can use the `returning: 'minimal'` option ([ref](https://supabase.io/docs/reference/javascript/insert#notes)), though that's probably not what you're looking for here.

**Reply by Nick-Mazuk** [reply]
Hmm… that's a good thought, though I don't think that's the issue (could be wrong). I added the policy to all operations by toggling "Advanced" and selecting "All".

<img width="439" alt="Screen Shot 2021-05-21 at 8 28 26 PM" src="https://user-images.githubusercontent.com/5659171/119213318-1a4a1480-ba73-11eb-8b49-1ae672ac6d39.png">

I can also confirm that the select statement should be working correctly. When plugging it into the SQL editor, it does return a table with my uid (truncated in the image below just in case it matters with security).

<img width="412" alt="Screen Shot 2021-05-21 at 8 31 43 PM" src="https://user-images.githubusercontent.com/5659171/119213395-8fb5e500-ba73-11eb-9529-86792af562be.png">

**Reply by soedirgo** [reply]
I tried using the same policy and it works fine, so I don't think the problem is in the policy definition. Not sure what could go wrong beyond that though...

**Comment by AndrewHows** [comment, marked as answer]
Have you enabled row-level security on the userRoles table? If you have, and haven't setup a policy to give users access, the policy on posts would fail due to not having the permission to run your select query.

**Reply by Nick-Mazuk** [reply]
That was it, thank you!

**Comment by ilyabo** [comment]
I had the same issue, but somehow just enabling RLS on the user roles table wasn't enough. I had to  grant `SELECT` permissions to `authenticated`:
```
GRANT SELECT ON TABLE public.user_roles TO authenticated;
GRANT SELECT ON TABLE public.user_roles TO anon;
```
